shell/platform/linux/main_linux.cc

```diff
@@ -14,22 +14,63 @@
 #include "flutter/shell/gpu/gpu_surface_gl.h"
 #include "flutter/shell/platform/linux/message_pump_glfw.h"
 #include "flutter/shell/platform/linux/platform_view_glfw.h"
+#include "flutter/shell/testing/test_runner.h"
 #include "flutter/shell/testing/testing.h"
+#include "flutter/sky/engine/public/web/Sky.h"
 
 namespace {
 
+// Checks whether the engine's main Dart isolate has no pending work.  If so,
+// then exit the given message loop.
+class ScriptCompletionTaskObserver : public base::MessageLoop::TaskObserver {
+ public:
+  ScriptCompletionTaskObserver(base::MessageLoop& message_loop)
+      : message_loop_(message_loop), prev_live_(false) {}
+
+  void WillProcessTask(const base::PendingTask& pending_task) override {}
+
+  void DidProcessTask(const base::PendingTask& pending_task) override {
+    shell::TestRunner& test_runner = shell::TestRunner::Shared();
+    bool live = test_runner.platform_view().engine().UIIsolateHasLivePorts();
+    if (prev_live_ && !live)
+      message_loop_.PostTask(FROM_HERE, message_loop_.QuitWhenIdleClosure());
+    prev_live_ = live;
+  }
+
+ private:
+  base::MessageLoop& message_loop_;
```

Comment:
You don't actually need to capture this pointer because it will be base::MessageLoop::current() in DidProcessTask